src/microbe_stage/CompoundCloudSystem.cs

```diff
@@ -1,210 +1,182 @@
+        if (isExiting)
+            return;
 
-        // Limit the rate at which the clouds are processed as they
-        // are a major performance sink
-        if (elapsed >= Settings.Instance.CloudUpdateInterval)
+        try
         {
-            UpdateCloudContents((float)elapsed);
-            elapsed = 0;
+            elapsed += delta;
+
+            if (elapsed >= Settings.Instance.CloudUpdateInterval)
+            {
+                elapsed = 0;
+                foreach (var cloudInstance in cloudInstances)
+                {
+                    cloudInstance.Call("update_texture");
+                }
+            }
         }
+        catch (Exception e)
+        {
+            GD.PrintErr($"Error in CompoundCloudSystem._Process: {e.Message}\n{e.StackTrace}");
+        }
+    }
+
+    private int GetCompoundID(CompoundDefinition compoundDefinition)
+    {
+        return (int)compoundDefinition.ID;
+    }
+
+    private int GetCompoundID(Compound compound)
+    {
+        return (int)compound;
+    }
+
+    public void SpawnCloud(Compound compound, float density, Vector3 worldPosition)
+    {
+        AddCloud(compound, density, worldPosition);
     }
 
-    /// <summary>
-    ///   Places specified amount of compound at position using interlocked operations for thread safety
-    /// </summary>
-    /// <returns>True when placing succeeded, false if out of range</returns>
     public bool AddCloud(Compound compound, float density, Vector3 worldPosition)
     {
-        // Find the target cloud //
-        foreach (var cloud in clouds)
+        int compoundID = GetCompoundID(compound);
+
+        foreach (var cloudInstance in cloudInstances)
         {
-            if (cloud.ContainsPosition(worldPosition, out int x, out int y))
+            bool contains = (bool)cloudInstance.Call("contains_position", worldPosition);
+            if (contains)
             {
-                // Within cloud
-
-                // Add if cloud handles this type
-                if (cloud.AddCloudInterlockedIfHandlesType(compound, x, y, density))
-                    return true;
+                var coords = (Vector2)cloudInstance.Call("convert_to_cloud_local", worldPosition);
+                cloudInstance.Call("add_cloud", compoundID, (int)coords.X, (int)coords.Y, density);
+                return true;
             }
         }
 
         return false;
     }
 
-    /// <summary>
-    ///   Takes compound at world position. This doesn't use locks or interlocked read so this is not thread safe
-    ///   unlike <see cref="AbsorbCompounds"/>, which is basically what should be used instead.
-    /// </summary>
-    /// <param name="compound">The compound type to take</param>
-    /// <param name="worldPosition">World position to take from</param>
-    /// <param name="fraction">The fraction of compound to take. Should be &lt;= 1</param>
     public float TakeCompound(Compound compound, Vector3 worldPosition, float fraction)
```

Comment:
I think this method needs to be re-thought out as this is called on many points for a circle for a single absorb operation, meaning that there's a ton of calls going between C# and C++ when absorbing. Also why did you remove the first part of the summary? And also you seemed to add a bunch of locks in the C++ code so it is less parallel than this C# code. You should instead consider trying to use atomic reads and writes in C++.